logicalplan/plan_test.go

```diff
@@ -132,3 +136,128 @@ func TestMatcherPropagation(t *testing.T) {
 		})
 	}
 }
+
+func TestDistributedExecution(t *testing.T) {
+	cases := []struct {
+		name     string
+		expr     string
+		expected string
+	}{
+		{
+			name: "sum-rate",
+			expr: `sum by (pod) (rate(http_requests_total[5m]))`,
+			expected: `
+sum by (pod) (
+  coalesce(
+    remote(sum by (pod) (rate(http_requests_total[5m]))),
+    remote(sum by (pod) (rate(http_requests_total[5m])))
+  )
+)`,
+		},
+		{
+			name: "avg",
+			expr: `avg by (pod) (http_requests_total)`,
+			expected: `
+avg by (pod) (
+  coalesce(
+    remote(http_requests_total),
+    remote(http_requests_total)
+  )
+)`,
```

Comment:
I am wondering for this query, do we want to do distributed execution or not? To me this seems better to just execute locally without querying remote engines. Since we are querying raw series, remote engines are just stores in this case

Reply:
I also thought about this, and the reason I thought it's beneficial to distribute the query is because PromQL is done over a limited number of steps. So by distributing the query we would "downsample" on the fly. When querying longer ranges, the number of samples going to the central querier will be much smaller than if we directly fetch chunks from storage. 

For example, the resolution for a 30d query is 10368s. So we will remove a significant amount of samples because remote engines will sample the data.

Reply:
Got it that makes sense to me. But one step further I am wondering if it is better to do this when retrieving series from stores. Like https://github.com/thanos-io/thanos/pull/4857

Reply:
The way it is currently implemented we would get the benefit of downsampling even if there is no function involved so it should be much better.

Reply:
> But for remote engines, all the raw samples are queried into memory, right? 

Yes, this is correct. However solving that at the storage level seems a bit more complicated and maybe something to tackle independently in the future.